dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/export/event/EventOperationParamsMapper.java

```diff
@@ -236,22 +265,115 @@ private OrganisationUnit validateOrgUnit(String orgUnitUid) throws BadRequestExc
     return orgUnit;
   }
 
-  private void validateUser(
-      User user, Program pr, ProgramStage ps, String orgUnitUid, OrganisationUnit orgUnit)
+  private void validateUser(User user, Program program, ProgramStage programStage)
       throws ForbiddenException {
-    if (pr != null && !user.isSuper() && !aclService.canDataRead(user, pr)) {
-      throw new ForbiddenException("User has no access to program: " + pr.getUid());
+    if (user.isSuper()) {
+      return;
+    }
+    if (program != null && !aclService.canDataRead(user, program)) {
+      throw new ForbiddenException("User has no access to program: " + program.getUid());
+    }
+
+    if (programStage != null && !aclService.canDataRead(user, programStage)) {
+      throw new ForbiddenException("User has no access to program stage: " + programStage.getUid());
+    }
+  }
+
+  /**
+   * Returns a list of all the org units the user has access to
+   *
+   * @param program the program the user wants to access to
+   * @param user the user to check the access of
+   * @param orgUnit parent org unit to get descendants/children of
+   * @param orgUnitDescendants function to retrieve org units, in case ou mode is descendants
+   * @return a list containing the user accessible organisation units
+   */
+  private List<OrganisationUnit> getUserAccessibleOrgUnits(
+      Program program,
+      User user,
+      OrganisationUnit orgUnit,
+      OrganisationUnitSelectionMode orgUnitMode,
+      Function<String, List<OrganisationUnit>> orgUnitDescendants) {
+
+    if (orgUnitMode == null && orgUnit != null) {
```

Comment:
Is this behavior documented?
How can this be achieved by a user? AFAIK DESCENDANTS is the default, would `orgUnitMode` be null when a request with `ouMode=` is made?

It might make it more readable if we set the `orgUnitMode` to SELECTED/ACCESSIBLE ideally before calling this method (your `getOrgUnitMode()` has the logic we need I think). We would then only need to return the switch expression.

Reply:
I'm not sure this is documented, but it's the conclusion I got after analyzing the code. There's [this ticket](https://dhis2.atlassian.net/browse/TECH-1588) to harmonize the endpoints when it comes to setting a default org unit mode

DESCENDANTS is set as the default but Spring is not picking it up (@enricocolasante knows more about this). Anyway, it does not make sense that DESCENDANTS is the default, but this will be tackled in the aforementioned ticket.

Regarding the change you requested, I think it makes sense to use the `getOrgUnitMode` method and have a clean switch expression in this one only.